test/Dialect/Calyx/round-trip.mlir

```diff
@@ -0,0 +1,47 @@
+// RUN: circt-opt %s -verify-diagnostics | FileCheck %s
```

Comment:
I think for these round trip tests, it is good to pipe through circt-opt again before piping to FileCheck. That way, you test that not only can circt-opt parse the IR as you wrote it, but also as the printer prints it.

Reply:
I did see this in the other test suites. However, we are taking in a MLIR-type input and outputting something that can be parsed by the current Calyx compiler, so running `circt-opt` twice won't work when we've changed, say, `calyx.component @MyComp` -> `component MyComp`.

Reply:
Hmmm I hadn't picked that up. So the MLIR printed form cannot be parsed by the MLIR parser? If that is the case, I think that may be an abuse of the printed form, and would recommend having an IR that round trips through `circt-opt | circt-opt`. The printed form of MLIR dialects is really only intended to be used by MLIR compiler developers to debug their compilers, and to write test cases. It's not really meant to be an interface to other systems.

To interface externally, you would add a [Translation](https://github.com/llvm/circt/tree/main/lib/Translation) like ExportCalyx to print it in a form the existing Calyx compiler understands. Even if that translation is going to print something very similar to the printed MLIR IR form, I think this is a pretty important separation of concerns. Down the road you may be adding things to the MLIR IR that need to be parsed/printed, but won't show up in the translation, for example.

Reply:
That makes sense, thanks for the explanation. I think a Translation to a Calyx-parsable format is the way to go. CC'ing @rachitnigam, @sampsyo to see if they have any input on this.

Reply:
Round-tripped for now.

Reply:
Yes, that definitely seems like the way to go! It resolves a conversation with @cgyurgyik about the seemingly-avoidable pain of using MLIR parsing to replicate the existing text format we invented & parsed separately. Sticking to a more "native MLIR" form for Calyx programs would be a better fit, and it's cool that there's precedent for using a Translation to emit an "external" text format.

That leaves, however, the question about how to convert _into_ the MLIR-native form from the Calyx-native form. I maintain that such a converter would not be too hard to hammer out by reusing the preexisting Calyx parser and in-memory IR. (Such a converter would have to remain in the Calyx project, not in the CIRCT repo.)

Reply:
Keeping the existing parser in Rust makes sense to me, but just FYI it's totally normal to also have Translations from an external format into MLIR as well. In CIRCT, for example, the FIRRTL project has a parser for their external format: https://github.com/llvm/circt/tree/main/lib/Dialect/FIRRTL/Import.

Reply:
Ah, neato, that could be cool too!